### PR TITLE
Homework2

### PR DESCRIPTION
**Pull request by marcindus**
Solution for Obiektowosc#2 zad#1, zad#2, zad#3, zad#4, Group: marcindus, adamvm

### CODE REVIEW

File: shm/Cargo.cpp [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    amount_ += amount;
+    return *this;
+}
+Cargo& Cargo::operator-=(size_t amount)
+{
+
+    if (amount_ >= amount)
+        amount_ -= amount;
+    else
+        throw AmountException("Not enough Cargo!");
```
Comment (WBigaj) [inline shm/Cargo.cpp]:
```suggestion
    if (amount_ >= amount) {
        amount_ -= amount;
    } else {
        throw AmountException("Not enough Cargo!");
    }
```

File: shm/Player.cpp [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+#include "ShmExceptions.hpp"
+
+void Player::Player::PayCrew(size_t money)
+{
+    if (money_ >= money)
+    {
+        money_ = money_ - money;
+    }
+    else
+        throw AmountException("Not enough money!");
```
Comment (WBigaj) [inline shm/Player.cpp]:
```suggestion
    else {
        throw AmountException("Not enough money!");
    }
```

File: shm/Ship.cpp [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        throw AmountException("Not enough Crew!");
+    return *this;
+}
+
+Ship& Ship::operator+=(size_t num)
+{
+    if ((crew_ + num) <= maxCrew_)
+        crew_ += num;
+    else
+        throw AmountException("Too much Crew!");
```
Comment (WBigaj) [inline shm/Ship.cpp]:
```suggestion
    if ((crew_ + num) <= maxCrew_) {
        crew_ += num;
    } else {
        throw AmountException("Too much Crew!");
    }
```

File: shm/Ship.cpp [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+{
+    name_ = name;
+}
+
+Ship& Ship::operator-=(size_t num)
+{
+    if (crew_ >= num)
+        crew_ -= num;
+    else
+        throw AmountException("Not enough Crew!");
```
Comment (WBigaj) [inline shm/Ship.cpp]:
```suggestion
    if (crew_ >= num) {
        crew_ -= num;
    } else {
        throw AmountException("Not enough Crew!");
    }
```

**Review by WBigaj — COMMENTED** [review]
Few suggestion from me how to improve code.

File: shm/Alcohol.cpp [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+Alcohol& Alcohol::operator+=(size_t amount)
+{
+    amount_ += amount;
+    return *this;
+}
+
+Alcohol& Alcohol::operator-=(size_t amount)
+{
+    amount_ -= amount;
```
Comment (niemarcin) [inline shm/Alcohol.cpp:38]: What if `amount` is greater than `amount_`?

File: shm/Alcohol.cpp [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    return *this;
+}
+
+bool Alcohol::operator==(const Cargo& cargo) const
+{
+    if (typeid(cargo) == typeid(Alcohol))
+    {
+        const auto* pAlcohol = dynamic_cast<const Alcohol*>(&cargo);
```
Comment (niemarcin) [inline shm/Alcohol.cpp:46]: Can't we use static_cast if we already checked that types are the same?

Comment (niemarcin) [reply shm/Alcohol.cpp:46]: Btw. I'm not sure but `auto pAlcohol` should be enough. Especially with static_cast.

File: shm/Fruit.cpp [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+Fruit& Fruit::operator+=(size_t amount)
+{
+    amount_ += amount;
+    return *this;
+}
+
+Fruit& Fruit::operator-=(size_t amount)
+{
+    amount_ -= amount;
```
Comment (niemarcin) [inline shm/Fruit.cpp:35]: What if `amount` is greater than `amount_`?

File: shm/Island.cpp [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+#include "Coordinates.hpp"
+
+Island::Island() : position_(Coordinates(0, 0))
+{
+}
+Island::Island(int x, int y) : position_(Coordinates(x, y))
+{
+}
+
+std::ostream& operator<<(std::ostream& os, Island& island)
```
Comment (niemarcin) [inline shm/Island.cpp:11]: Assume that `Island` is part of DLL which you provide to your customer. If it was defined as a `friend` of Island class, then "end user" who has DLL and .hpp file would know that there is << operator for Island class. In your case user won't have any idea about it.

File: shm/Island.hpp [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+    Island();
+    Island(int x, int y);
+
+    Coordinates getPosition() { return position_; }
+
+private:
+    Coordinates position_;
+};
+
+bool operator==(Island& lhs, Island& rhs);
```
Comment (niemarcin) [inline shm/Island.hpp:19]: Why it's not a part of Island class? Could it be defined as static method inside the class?

File: shm/Map.cpp [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+#include <iostream>
+
+#include "Map.hpp"
+#include "ShmExceptions.hpp"
+
+Map::Map() : mapXsize(defaultXsize), mapYsize(defaultYsize)
+{
+    randomIslandGenerator(defaultIslandsAmount);
+}
+Map::Map(int sizeY, int sizeX, int islandsAmount) : mapXsize(sizeY), mapYsize(sizeX)
```
Comment (niemarcin) [inline shm/Map.cpp:11]: mapXsize and mapYsize are size_t, while sizeX and sizeY are int type

btw. you initialize mapXsize with sizeY and mapYsize with sizeX values - x with y and y with x.

File: shm/Map.cpp [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+{
+    if (islandsAmount <= (sizeY * sizeX))
```
Comment (niemarcin) [inline shm/Map.cpp:13]: what if sizeY or sizeX is negative value?

File: shm/Map.cpp [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+        randomIslandGenerator(islandsAmount);
+    }
+    else
+    {
+        throw AmountException("Too many Islands!");
+        Map();
+    }
+}
+
+Island Map::getIsland(const Coordinates& corr)
```
Comment (niemarcin) [inline shm/Map.cpp:24]: Could it be const method?

File: shm/Map.hpp [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+#include <algorithm>
+#include <iostream>
+#include <random>
+#include <vector>
+
+#include "Island.hpp"
+
+constexpr size_t defaultXsize = 10;
+constexpr size_t defaultYsize = 11;
+constexpr int defaultIslandsAmount = 10;
```
Comment (niemarcin) [inline shm/Map.hpp:12]: Can "amount" be <0 ? I mean - why it's an `int`?

**Review by niemarcin — APPROVED** [review]
Few things to consider ;)

**Review by MaciejSalamonski — APPROVED** [review]
Nice work :)

File: shm/Alcohol.cpp [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+{
+}
+std::string Alcohol::GetName() const
+{
+    return name_;
+};
+
+size_t Alcohol::GetPrice() const
+{
+    return Alcohol::GetBasePrice() * Alcohol::GetAmount();
```
Comment (nauka-programowania-MA) [inline shm/Alcohol.cpp:17]: Why `Alcochol::` prefix?

File: shm/Alcohol.cpp (hunk 6 above)
Comment (nauka-programowania-MA) [reply shm/Alcohol.cpp:46]: prefer use static_cast over unsafe dynamic_cast

File: shm/Cargo.cpp [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+Cargo& Cargo::operator-=(size_t amount)
+{
+
+    if (amount_ >= amount)
+    {
+        amount_ -= amount;
+    }
+    else
+    {
+        throw AmountException("Not enough Cargo!");
```
Comment (nauka-programowania-MA) [inline shm/Cargo.cpp:25]: are you handling exceptions?

File: shm/Fruit.cpp [hunk 16]
```diff
@@ -0,0 +1,88 @@
...
+    return *this;
+}
+
+bool Fruit::operator==(const Cargo& cargo) const
+{
+    if (typeid(cargo) == typeid(Fruit))
+    {
+        const auto* pFruit = dynamic_cast<const Fruit*>(&cargo);
```
Comment (nauka-programowania-MA) [inline shm/Fruit.cpp:43]: staitc_cast

File: shm/Map.cpp (hunk 10 above)
Comment (nauka-programowania-MA) [reply shm/Map.cpp:11]: +1

File: shm/Map.cpp (hunk 11 above)
Comment (nauka-programowania-MA) [reply shm/Map.cpp:13]: +1

File: shm/Map.hpp (hunk 13 above)
Comment (nauka-programowania-MA) [reply shm/Map.hpp:12]: +1

**Review by nauka-programowania-MA — APPROVED** [review]

**Comment by nauka-programowania-MA** [comment]
{
"requirement_ids": [2568, 2568, 2568, 2571, 2571, 2571],
"points": [5, 5, 5, 8, 8, 8],
"users": ["marcindus", "adamvm"]
}

**Pull request closed without merging**